### PR TITLE
[ADD] Customizable templates for sonata page controller

### PR DESCRIPTION
**Pull request by Th3Mouk**
Add flexibility for extend templates (available since 2.3.10)

### CODE REVIEW

File: Controller/PageAdminController.php [hunk 1]
```diff
@@ -94,7 +94,7 @@ public function treeAction()
 
         $this->get('twig')->getExtension('form')->renderer->setTheme($formView, $this->admin->getFilterTheme());
 
-        return $this->render('SonataPageBundle:PageAdmin:tree.html.twig', array(
+        return $this->render($this->container->getParameter('sonata.page.templates.admin')['tree'], array(
```
Comment (rande) [inline Controller/PageAdminController.php]: This is not compatible with php 5.3 :facepunch:

Comment (Th3Mouk) [inline Controller/PageAdminController.php]: damn i've forget it, i will correct and switch my env

**Comment by Th3Mouk** [comment]
All green <3

File: Controller/PageAdminController.php [hunk 2]
```diff
@@ -240,10 +246,22 @@ public function composeContainerShowAction()
...
             'page'          => $block->getPage(),
         ));
     }
+
+    /**
+     * Function to retrieve customizable templates.
+     *
+     * @return mixed
+     */
+    public function retrieveAdminTemplates()
```
Comment (rande) [inline Controller/PageAdminController.php]: must be protected

**Comment by Soullivaneuh** [comment]
As a new feature, this should be on next minor version, not patch.

Please make another PR on `master`.

**Comment by Soullivaneuh** [comment]
Forget my comment, read it to fast.

**Comment by Th3Mouk** [comment]
@Soullivaneuh do yo needs additional work ?

File: Controller/PageAdminController.php [hunk 3]
```diff
@@ -94,7 +94,9 @@ public function treeAction()
 
         $this->get('twig')->getExtension('form')->renderer->setTheme($formView, $this->admin->getFilterTheme());
 
-        return $this->render('SonataPageBundle:PageAdmin:tree.html.twig', array(
+        $templates = $this->retrieveAdminTemplates();
+
+        return $this->render($templates['tree'], array(
```
Comment (Soullivaneuh) [inline Controller/PageAdminController.php]: Whit not using `$this->admin->getTemplate('tree')` directly instead of getting all admin templates?

Comment (Th3Mouk) [inline Controller/PageAdminController.php]: seem logic i'll work on today, the first problem was related with PHP 5.3 compatibility.
Is the `retrieveAdminTemplates()` function in its place ?

Comment (Soullivaneuh) [inline Controller/PageAdminController.php]: What is the relation with PHP 5.3?

The line `$templates = $this->retrieveAdminTemplates();` is not needed anymore if you use the following code:

```php
return $this->render($this->admin->getTemplate('tree'), array());
```

Comment (Th3Mouk) [inline Controller/PageAdminController.php]: Because it's an array of parameters so the
```
$this->container->getParameter('sonata.page.templates.admin')['tree']
```
isn't available, that's why i made it in 2 step.

I can't
```
$this->admin->getTemplate('tree')
```
Because this parameters are related to SonataPageBundle and not SonataAdminBundle, the config node is templates_admin.

I don't know how to add this entries to SonataAdmin's `getTemplates()` collection from this bundle's file `SonataPageExtension.php`

Comment (Soullivaneuh) [inline Controller/PageAdminController.php]: Maybe something could be done here: https://github.com/sonata-project/SonataPageBundle/blob/42e550a0b2b561520ed233094b633a0d1b4706d2/Resources/config/admin.xml#L64-L68

Comment (Th3Mouk) [inline Controller/PageAdminController.php]: seems best way too :+1: 
so customizing templates don't will be from `configuration.php` like SonataAdmin's templates but from parameters.

**Comment by Th3Mouk** [comment]
@Soullivaneuh simply and clear thanks

File: Resources/config/admin.xml [hunk 4]
```diff
@@ -28,6 +28,13 @@
         <parameter key="sonata.page.admin.snapshot.class">Sonata\PageBundle\Admin\SnapshotAdmin</parameter>
         <parameter key="sonata.page.admin.snapshot.controller">SonataPageBundle:SnapshotAdmin</parameter>
         <parameter key="sonata.page.admin.snapshot.translation_domain">%sonata.page.admin.page.translation_domain%</parameter>
+
+        <!-- TEMPLATES -->
+        <parameter key="sonata.page.admin.template.list">SonataPageBundle:PageAdmin:list.html.twig</parameter>
```
Comment (Soullivaneuh) [inline Resources/config/admin.xml]: Parameters are not needed because used only for `setTemplates` method call IMO.

File: Resources/doc/reference/advanced_usage.rst [hunk 5]
```diff
@@ -38,3 +38,15 @@ core routes):
     sonata_page:
         page_defaults:
             homepage: {decorate: false} # disable decoration for homepage route_name
+
+Templates customization
+-----------------------
+
+You can also customize/overwriting the templates of the bundle.
```
Comment (Soullivaneuh) [inline Resources/doc/reference/advanced_usage.rst]: Template override is done with `app/Resources` trick, not need of that.

Comment (Th3Mouk) [inline Resources/doc/reference/advanced_usage.rst]: Sometimes you prefer use inheritance and extends original template to ease vendor updates and only modify part of templates. Should I delete this part ?

Comment (Soullivaneuh) [inline Resources/doc/reference/advanced_usage.rst]: Yes, other sonata bundle works with the described method AFAIK.

**Comment by Th3Mouk** [comment]
@Soullivaneuh @OskarStark any advises ?

**Comment by OskarStark** [comment]
@Th3Mouk please remove the Parameters as mentioned by @Soullivaneuh and squash your commits!

Thank you! :+1:

**Comment by Soullivaneuh** [comment]
@OskarStark As I said on https://github.com/sonata-project/SonataPageBundle/pull/528#issuecomment-139569214:

>

**Comment by Soullivaneuh** [comment]
Sorry, bad manipulation.

**Comment by Th3Mouk** [comment]
@OskarStark i need this parameters because it's more easy to overwrite a template passing by configuration's parameters, compared to a pass compilator or rewriting a service, and inject it on an other service etc .. This solution seem the best solution for overwriting template IMO

**Comment by Soullivaneuh** [comment]
@Th3Mouk Should no be overridden like this. Should be on sonata config IMO.

BTW, this method is not recommended anymore by Symfony.

**Comment by Th3Mouk** [comment]
@Soullivaneuh OK thanks, i didn't know that isn't recommended

**Comment by Th3Mouk** [comment]
@Soullivaneuh any other comment ?

File: Admin/SharedBlockAdmin.php [hunk 6]
```diff
@@ -88,15 +88,15 @@ protected function configureFormFields(FormMapper $formMapper)
     }
 
     /**
-     * {@inheritDoc)
+     * {@inheritDoc).
```
Comment (core23) [inline Admin/SharedBlockAdmin.php]: No point and complete lowercase

Comment (core23) [inline Admin/SharedBlockAdmin.php]: Maybe create a seperate PR for the phpdoc changes as they are not relevant for this feature

Comment (Th3Mouk) [inline Admin/SharedBlockAdmin.php]: they're are due to `make cs` command

Comment (Th3Mouk) [inline Admin/SharedBlockAdmin.php]: not relative to this PR

Comment (core23) [inline Admin/SharedBlockAdmin.php]: Please revert these changes. Only checkin files that essential for the PR

File: Admin/SharedBlockAdmin.php [hunk 7]
```diff
@@ -88,15 +88,15 @@ protected function configureFormFields(FormMapper $formMapper)
...
      */
     public function createQuery($context = 'list')
     {
         $query = parent::createQuery($context);
 
         // Filter on blocks without page and parents
-        $query->andWhere($query->expr()->isNull($query->getRootAlias() . '.page'));
-        $query->andWhere($query->expr()->isNull($query->getRootAlias() . '.parent'));
+        $query->andWhere($query->expr()->isNull($query->getRootAlias().'.page'));
+        $query->andWhere($query->expr()->isNull($query->getRootAlias().'.parent'));
```
Comment (core23) [inline Admin/SharedBlockAdmin.php]: Same here. Please revert as it is not relevant for this PR.

Comment (Th3Mouk) [inline Admin/SharedBlockAdmin.php]: OK !

**Comment by Th3Mouk** [comment]
ping @Soullivaneuh

**Comment by OskarStark** [comment]
@Th3Mouk can you please squash your commits?

**Comment by Th3Mouk** [comment]
@OskarStark :+1:

**Comment by OskarStark** [comment]
Thank you, i will have a look at the travis problem

**Comment by core23** [comment]
Please rebase @Th3Mouk Travis should be okay

**Comment by Th3Mouk** [comment]
@core23 the fail is not relative to this PR, commit b05131527fce96abee1dccab22f3a271f55124e9 have the same problem

**Comment by core23** [comment]
Okay for me

**Comment by OskarStark** [comment]
Thank you @Th3Mouk can you please port this commit to the master branch?
